### ISSUE TITLE
detekt-config.yml is ignored when running type resolution enabled tasks in a KMM multiplatform projects

### ISSUE THREAD

**Issue by dimsuz** [detekt/detekt#4380]
## Expected Behavior
Running either `detektAndroidDebug` or `detektMetadataMain` uses settings from `detekt-config.yaml`

## Observed Behavior
On a KMM project when `detektAndroidDebug` task is executed, default config options are used, ignoring settings in the configured yaml file.

If instead I run `detekt` task (without type resolution), then Detekt repspects config file and finds way more (or less, depending on the config) issues.

## Steps to Reproduce
1. Open attached sample project
2. Run `detektAndroidDebug` task: `./gradlew detektAndroidDebug`
3. Observe that build fails with an `MemberNameEqualsClassName` issue, while it is being explicitely disabled in `detekt-config.yaml`
4. Run `detekt` task. 
5. Observe that 13 errors were found (in line with config), and it doesn't include `MemberNameEqualsClassName` which is not active

## Your Environment

* Version of detekt used: `1.19.0`
* Version of Gradle used (if applicable): `7.0.2`

## Sample project
[detekt-tr-no-config-bug.zip](https://github.com/detekt/detekt/files/7735735/detekt-tr-no-config-bug.zip)

**Comment by dimsuz** [comment]
Tried to debug this, inserted a lot of logs (I'm also new to the detekt codebase, so don't know where to look right away).

I want to document my (incomplete) findings.

Sample project has the following file config:

```kotlin
tasks.withType<io.gitlab.arturbosch.detekt.Detekt>() {
  config.setFrom(files(rootProject.file("detekt-config.yml")))
}
```

These get turned to a cli-arguments and eventually passed by gradle plugin to a `detekt-cli`.

The problem is that for some reason that same line above when read from `DetektExtension` within [DetektMultiplatform.kt](https://github.com/detekt/detekt/blob/3e195d63b4f79d06c4337a0ee6cb8a8a31ee783a/detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/internal/DetektMultiplatform.kt#L67) which calls `Project.registerDetektTask()` to create `detektAndroidDebug`, causes `extension.config.files` to be **EMPTY**, but when that **same** extension object is passed during creation of `detekt` task then `extension.config.files` object contains full path to a `detekt-config.yml`. Very weird!

So for `detekt` task cli gets `--config /path/to/yml` while for `detektAndroidDebug` task no `--config` option is created.

At this point I feel like I don't have enough knowledge of gradle internals to find out why that happens.

Also it seems that `detekt` task creation originates from gradle itself bypassing `DetektMultiplatform.kt` functions (they are not in the stacktrace if I print one), while type-resolution tasks are being created from `DetektMultiplatform.kt` (it is in the stack trace).

**Comment by dimsuz** [comment]
I also think that if one adds some `yml` file and configures the tests below to check for `--config` option, they will fail:

```kotlin
it("configures detekt task with type resolution") {
    gradleRunner.runTasksAndCheckResult(":shared:detektAndroidDebug") {
        assertThat(it.output).containsPattern("""--baseline \S*[/\\]detekt-baseline-debug.xml """)
        !!!! ADD THIS, TEST SHOULD FAIL:
        assertThat(it.output).containsPattern("""--config /path/to/some/yaml""")
        assertDetektWithClasspath(it)
    }
}
```

Here are the tests in question: [source link](https://github.com/detekt/detekt/blob/78e6d83010a2f76077005430cee13a08ebcf0699/detekt-gradle-plugin/src/test/kotlin/io/gitlab/arturbosch/detekt/DetektMultiplatformSpec.kt#L187-L196)

But I didn't figure right away how to add some yaml file into a test environments without risking breaking other tests in this file.

**Comment by BraisGabin** [comment]
I did some research by my own about this. I see that if you set the config file in the default path `config/detekt/detekt.yml` it works. And if you use this syntax:

```kotlin
detekt {
  config = files("detekt.yml")
}
```

It works too.

But I wasn't able to make it work with:

```kotlin
tasks.withType<io.gitlab.arturbosch.detekt.Detekt>() {
  config.setFrom(files(rootProject.file("detekt-config.yml")))
}
```

I'm getting this error message:

```
Could not compile build file '/private/var/folders/_9/ntvzf7jd6y30cr9rxs154vl996qbgx/T/applyPlugin12537619203849001420/shared/build.gradle'.
> startup failed:
  build file '/private/var/folders/_9/ntvzf7jd6y30cr9rxs154vl996qbgx/T/applyPlugin12537619203849001420/shared/build.gradle': 19: Unexpected input: '{' @ line 19, column 78.
     b.arturbosch.detekt.Detekt>() {
                                   ^
  
  1 error
```

(I did all the tests using the current Unit tests)

**Comment by cortinico** [comment]
I spent some time investigating this. Turns out that your configuration is not correct @dimsuz. A correct configuration would be:

```kotlin
allprojects {
    apply(plugin = "io.gitlab.arturbosch.detekt")
    detekt {
        parallel = false
        config.setFrom(files(rootProject.file("detekt-config.yml")))
    }
    tasks.withType<io.gitlab.arturbosch.detekt.Detekt>().configureEach {
        jvmTarget = "11"
        reports {
            xml.required.set(false)
            html.required.set(false)
            txt.required.set(false)
            sarif.required.set(false)
        }
        exclude("**/resources/**")
        exclude("**/build/**")
    }
}
```

Or you can wrap your `tasks.withType<>{}` inside a `project.afterEvaluate` (which is not great). 

So you had a couple of problems:

1. You were not using the `detekt{}` extensions. This means that default value got provided (for a config, is an empty path). Those were essentially overriding the values you provided.

2. You were setting the source for all the Detekt tasks to all the folders. I'm not sure why you were doing this but essentially that goes against the idea of offering multiple detekt tasks, one for every source set.

The fact that:

```
    tasks.withType<io.gitlab.arturbosch.detekt.Detekt>().configureEach {
        config.setFrom(...)
    }
```

is ignored can be considered a bug but I'm unsure if we want to address this, as several of the public fields of the `Detekt` tasks have a similar behavior.

**Comment by dimsuz** [comment]
Thank you very much for looking into this! For `1.` I have migrated my previous "configure root-project task to run on all .kt files" config to "run in each module config" and so I must've inherited this incorrect config. As for `2.` I must've done it thinking that somehow it doesn't work by default, although I've then used per-source set tasks, didn't notice that this is illogical :)

I guess the main thing that caused me to think this is a bug is that my config worked perfectly fine without type resolution enabled and didn't work with it. Maybe it's due to that possible bug you mention lastly.

So I guess this can be closed then, thanks again!
